[addons] add new IAddonInstanceHandler

AlwinEsch opened the pull request:
<!--- Provide a general summary of your change in the Title above -->

## Description
This class becomes in future used as parent on the various binary
addon instances as alternative to the CAddonDll who can no more
used as parent.

Currently this class include only the minimum but predefine the
coming multiple instance support.

Also is screensaver changed to new way and nearly in the style on end.
<!--- Describe your change in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

AlwinEsch commented:
Needed changes for this request are done. @Paxxi @Montellese @ksooo is this style OK?

ksooo commented on xbmc/addons/AddonInstanceHandler.cpp:
```diff
@@ -0,0 +1,57 @@
+
+#include "AddonInstanceHandler.h"
+
+#include "utils/log.h"
```
No functionality from utils/logs.h used in this ccp file, as far as I can see. Could be removed, imo.

ksooo commented on xbmc/addons/AddonInstanceHandler.cpp:
```diff
@@ -0,0 +1,57 @@
+{
+  if (m_addon && m_addon->CreateInstance(m_type, m_instanceId, instance, m_parentInstance) == ADDON_STATUS_OK)
+    return true;
+  return false;
```
Why not just <code>return (m_addon && m_addon->CreateInstance(m_type, m_instanceId, instance, m_parentInstance) == ADDON_STATUS_OK)</code>?

ksooo commented on xbmc/addons/AddonInstanceHandler.h:
```diff
@@ -0,0 +1,57 @@
+    std::string Icon() const { return m_addon->Icon(); }
+    std::string Path() const { return m_addon->Path(); }
+    std::string Profile() const { return m_addon->Profile(); }
+    AddonVersion Version() const { return m_addon->Version(); }
```
In the cpp file you are checking m_addon for nullptr before accessing it. Why not in the header file?

AlwinEsch committed:
[addons] add new IAddonInstanceHandler

This class becomes in future used as parent on the various binary
addon instances as alternative to the CAddonDll who can no more
used as parent.

Currently this class include only the minimum but predefine the
coming multiple instance support.

AlwinEsch commented:
@ksooo have added a check to the functions on header, further is a new commit there who translate the addon status to a string and becomes used on IAddonInstanceHandler to bring a error.

The `m_addon` as nullptr brings now a LOGFATAL, normally should this never happen.

AlwinEsch committed:
[addons] add addon status translator
This add a addon header function to translate a addon status value
to a string.

Further becomes this used in the IAddonInstanceHandler to get error.

ksooo commented:
@AlwinEsch From my point of view okay to merge now. cannot really judge on the screensaver changes but I guess this is good to go as well.

ksooo commented:
jenkins build this please

AlwinEsch committed:
[screensaver] change screensaver to multiple instance way

This change all screensaver parts to support the coming multiple instance
way.

Multiple instances are currently still not possible but on Screensaver
must be nearly nothing changed anymore.

One part who next bigger check is the Python screensaver addon, his
parts are removed from CScreenSaver and handled direct in CApplication.

AlwinEsch commented:
jenkins build this please
